### PR TITLE
#13 deadlock detection (now also for chain-signals)

### PR DESCRIPTION
**Pull request by alimasigna**
added deadlock_analyser class for recursion method
added result class mainly for debug reasons but also to save the result if a deadlock was found
added getProtected() method in Node class (because it was missing)
added deadlock_analyser init in app.class
increased size of readable blueprint in Decoder class for testing purpose
added getSections() method in Graph class to get all Sections for deadlock_analyser
added empty constructor for Matrix class (not necessary yet, probably later for UI but not bad to have at all)

### CODE REVIEW

File: app/src/main/java/factorio/train/analyser/analyser/Deadlock_Analyser.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
+package factorio.train.analyser.analyser;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import factorio.train.analyser.graph.Graph;
+import factorio.train.analyser.graph.Section;
+import factorio.train.analyser.graph.Node;
+import java.util.Map.Entry;
+
+public class Deadlock_Analyser {
```
Comment (TheHKT) [inline app/src/main/java/factorio/train/analyser/analyser/Deadlock_Analyser.java]: Please rename the class DeadlockAnalyser. We user camelCase only here!!

File: app/src/main/java/factorio/train/analyser/analyser/Deadlock_Analyser.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    public ArrayList<Result> deadlockCheck() {
+        if(graph != null){
+            ArrayList<Result> result = deadlockDetection(graph.getSections());
+            return result;
+        }else{
+            return null;
+        }
+    }
+
+    private ArrayList<Result> deadlockDetection(ArrayList<Section> sections) {
```
Comment (TheHKT) [inline app/src/main/java/factorio/train/analyser/analyser/Deadlock_Analyser.java]: pls add Javadoc comments to all the methods and fields. I recommend using copilot.

File: app/src/main/java/factorio/train/analyser/analyser/Deadlock_Analyser.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                            if ((nextNode != predecessor && !findPair(resultYet.getChainSignalsVisited(), nextNode, nodeInSection)) || nextNode.equals(predecessor) && nodeInSection.getIsInput() && !findPair(resultYet.getChainSignalsVisited(), nextNode, nodeInSection) ||!(!inList(nodeInSection.getProtectedFrom(), predecessor) && nextNode.equals(predecessor)) && !findPair(resultYet.getChainSignalsVisited(), nextNode, nodeInSection)) {
+                                HashMap newPair = new HashMap<Node, Node>();
+                                newPair.put(nodeInSection, nextNode);
+                                resultYet.addToChainSignalsVisited(newPair);
+                                recursion(nextNode, resultYet, nodeInSection);
+                            }
+                        }
+                    }
+                }
+                else { //no protection, just go to the next node
```
Comment (TheHKT) [inline app/src/main/java/factorio/train/analyser/analyser/Deadlock_Analyser.java]: move the "else" { higher it should look like "} else {"

File: app/src/main/java/factorio/train/analyser/analyser/Result.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+
+import factorio.train.analyser.graph.Node;
+
+public class Result {
+    
+    private ArrayList<Node> deadlockPath;
+    private ArrayList<HashMap<Node, Node>> chainSignalsVisited;
+    private boolean result;
```
Comment (TheHKT) [inline app/src/main/java/factorio/train/analyser/analyser/Result.java]: rename result to something like "isDeadlock" or "deadlockFound"

File: app/src/main/java/factorio/train/analyser/Matrix.java [hunk 5]
```diff
@@ -33,6 +33,8 @@ public Matrix(String encodedJson){
         setMatrix();
     }
 
+    public Matrix() {}
```
Comment (TheHKT) [inline app/src/main/java/factorio/train/analyser/Matrix.java:36]: Why do you need that constructor?

**Review by TheHKT — CHANGES_REQUESTED** [review]

File: app/src/main/java/factorio/train/analyser/Matrix.java (hunk 5 above)
Comment (alimasigna) [reply app/src/main/java/factorio/train/analyser/Matrix.java:36]: later probably for UI stuff. But isnt it common case in java to have an construktor anyway?

Comment (TheHKT) [reply app/src/main/java/factorio/train/analyser/Matrix.java:36]: well we already have a constructor tho. i mean its ok but if we dont use that constructor i guess we remove it

**Review by TheHKT — APPROVED** [review]

**Pull request merged**